Makefile

```diff
@@ -1,92 +1,145 @@
-        $(wildcard app/*.c) \
-        $(USB_SRCS)
+ifeq ($(NXDK_DIR),)
+NXDK_DIR = $(shell pwd)
+endif
+
+ifeq ($(XBE_TITLE),)
+XBE_TITLE = nxdk_app
+endif
+
+ifeq ($(OUTPUT_DIR),)
+OUTPUT_DIR = bin
+endif
+
+UNAME_S := $(shell uname -s)
+ifeq ($(UNAME_S),Linux)
+LD           = x86_64-w64-mingw32-ld
+CC           = clang
+CXX          = clang++
+endif
+ifeq ($(UNAME_S),Darwin)
+LD           = /usr/local/opt/llvm/bin/lld -flavor link
+CC           = /usr/local/opt/llvm/bin/clang
+CXX          = /usr/local/opt/llvm/bin/clang++
+endif
+
+TARGET       = $(OUTPUT_DIR)/default.xbe
+CGC          = wine $(NXDK_DIR)/tools/cg/cgc.exe
+CXBE         = $(NXDK_DIR)/tools/cxbe/cxbe
+VP20COMPILER = $(NXDK_DIR)/tools/vp20compiler/vp20compiler
+FP20COMPILER = $(NXDK_DIR)/tools/fp20compiler/fp20compiler
+EXTRACT_XISO = $(NXDK_DIR)/tools/extract-xiso/extract-xiso
+TOOLS        = cxbe vp20compiler fp20compiler extract-xiso
+CFLAGS       = -target i386-pc-win32 -march=pentium3 \
+               -ffreestanding -nostdlib -fno-builtin -fno-exceptions \
+               -I$(NXDK_DIR)/lib -I$(NXDK_DIR)/lib/xboxrt \
+               -Wno-ignored-attributes
+
+ifeq ($(DEBUG),y)
+CFLAGS += -g
+endif
+
+include $(NXDK_DIR)/lib/Makefile
 OBJS = $(SRCS:.c=.obj)
 
-bin/default.xbe: app/main.exe
-	mkdir -p bin
-	$(CXBE) -OUT:bin/default.xbe -TITLE:0ldskoo1 app/main.exe
-
-app/main.exe: $(OBJS)
-	$(LLDLINK) -subsystem:windows -dll -out:'$@' -entry:XboxCRT xboxkrnl/libxboxkrnl.lib $(OBJS)
-
-%.obj: %.c ${INCLUDES}
-	$(CC) $(CFLAGS) -c -o '$@' '$<'
-
-app/vs.vp: app/vs.cg
-	$(CGC) -profile vp20 -o '$@' '$<'
-
-app/ps.fp: app/ps.cg
-	$(CGC) -profile fp20 -o '$@' '$<'
-
-%.inl: %.vp
-	$(VP20COMPILER) '$<' > '$@'
-
-%.inl: %.fp
-	$(FP20COMPILER) '$<' > '$@'
+ifneq ($(GEN_XISO),)
+TARGET += $(GEN_XISO)
+endif
+
+V = 0
+VE_0 := @
+VE_1 :=
+VE = $(VE_$(V))
+
+ifeq ($(V),1)
+STDOUT_TO_NULL=
+else
+STDOUT_TO_NULL=>/dev/null
+endif
+
+DEPS := $(SRCS:.c=.c.d)
+
+all: $(TARGET)
+
+$(OUTPUT_DIR)/default.xbe: $(OUTPUT_DIR) main.exe
+	@echo "[ CXBE     ] $@"
+	$(VE)$(CXBE) -OUT:$@ -TITLE:$(XBE_TITLE) $^ $(STDOUT_TO_NULL)
+
+$(OUTPUT_DIR):
+	@mkdir -p $(OUTPUT_DIR);
+
+ifneq ($(GEN_XISO),)
+$(GEN_XISO): $(OUTPUT_DIR)/default.xbe
+	@echo "[ XISO     ] $@"
+	$(VE) $(EXTRACT_XISO) -c $(OUTPUT_DIR) $(XISO_FLAGS) $@ $(STDOUT_TO_NULL)
+endif
+
+main.exe: $(OBJS) $(NXDK_DIR)/lib/xboxkrnl/libxboxkrnl.lib
+	@echo "[ LD       ] $@"
+ifeq ($(UNAME_S),Linux)
+	$(VE) $(LD) -m i386pe -shared --entry=_XboxCRT -o $@ $^
+endif
+ifeq ($(UNAME_S),Darwin)
+	$(VE) $(LD) -subsystem:windows -dll -out:'$@' -entry:XboxCRT $^
+endif
+
+%.obj: %.c
+	@echo "[ CC       ] $@"
+	$(VE) $(CC) $(CFLAGS) -c -o '$@' '$<'
+
+%.c.d: %.c
+	@echo "[ DEP      ] $@"
+	$(VE) set -e; rm -f $@; \
+	$(CC) -M -MM -MG -MT '$*.obj' -MF $@ $(CFLAGS) $<; \
+	echo "\n$@ : $^\n" >> $@
+
+%.inl: %.vs.cg
+	@echo "[ CG       ] $@"
+	$(VE) $(CGC) -profile vp20 -o $@.$$$$ $< $(STDOUT_TO_NULL) && \
+	$(VP20COMPILER) $@.$$$$ > $@ && \
+	rm -rf $@.$$$$
```

Comment:
Possibly. The $@.$$$$ will generate a new file each time it runs bc $$$$ gets translated to the pid. We could possibly do a string substitution to keep better track of the intermediates.